src/Evolution/Systems/RelativisticEuler/Valencia/Characteristics.cpp

```diff
@@ -87,20 +92,324 @@ std::array<DataVector, Dim + 2> characteristic_speeds(
+
+  const double h_w_vn_c_over_d = specific_enthalpy_times_lorentz_factor *
+                                 normal_velocity * sound_speed_over_d;
+
+  Matrix result(Dim + 2, Dim + 2);
+
+  // set eigenvector with eigenvalue = lapse * Lambda_minus - normal_shift
+  size_t eigenvector_id = 0;
+
+  result(0, eigenvector_id) = 1.0;
+  result(1, eigenvector_id) = h_w_minus_one - h_w_vn_c_over_d;
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = specific_enthalpy_times_lorentz_factor *
+                                    (spatial_velocity_oneform.get(i) -
+                                     sound_speed_over_d * unit_normal.get(i));
+  }
+
+  // set eigenvector with eigenvalue = lapse * Lambda_plus - normal_shift
+  eigenvector_id = Dim + 1;
+
+  result(0, eigenvector_id) = 1.0;
+  result(1, eigenvector_id) = h_w_minus_one + h_w_vn_c_over_d;
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = specific_enthalpy_times_lorentz_factor *
+                                    (spatial_velocity_oneform.get(i) +
+                                     sound_speed_over_d * unit_normal.get(i));
+  }
+
+  // set eigenvector with eigenvalue = lapse * normal_velocity - normal_shift
+  eigenvector_id = Dim;
+
+  result(0, eigenvector_id) = get(kappa_over_density);
+  result(1, eigenvector_id) =
+      get(kappa_over_density) * h_w_minus_one -
+      specific_enthalpy_times_lorentz_factor * get(sound_speed_squared);
+
+  const double prefactor = specific_enthalpy_times_lorentz_factor *
+                           (get(kappa_over_density) - get(sound_speed_squared));
+  for (size_t i = 0; i < Dim; ++i) {
+    result(i + 2, eigenvector_id) = prefactor * spatial_velocity_oneform.get(i);
+  }
+
+  // if Dim > 1 set degenerate eigenvectors
+  if (Dim > 1) {
+    const double two_h_w_minus_one =
+        h_w_minus_one + specific_enthalpy_times_lorentz_factor;
+    const auto unit_tangent_oneform =
+        orthonormal_oneform(unit_normal, inv_spatial_metric);
+    const auto set_degenerate_eigenvector =
+        [
+          &result, &lorentz_factor, &spatial_velocity,
+          &spatial_velocity_oneform, &two_h_w_minus_one, &specific_enthalpy
+        ](const size_t vector_id,
+          const tnsr::i<double, Dim>& tangent_oneform) noexcept {
+      result(0, vector_id) =
+          get(lorentz_factor) *
+          get(dot_product(tangent_oneform, spatial_velocity));
+      result(1, vector_id) = two_h_w_minus_one * result(0, vector_id);
+      const double local_prefactor =
+          2.0 * get(lorentz_factor) * result(0, vector_id);
+      for (size_t i = 0; i < Dim; ++i) {
+        result(i + 2, vector_id) =
+            get(specific_enthalpy) *
+            (tangent_oneform.get(i) +
+             local_prefactor * spatial_velocity_oneform.get(i));
+      }
+    };
+    set_degenerate_eigenvector(1, unit_tangent_oneform);
+
+    if (Dim > 2) {
+      // orthonormal_form for two forms is defined for 3-d only
```

Comment:
`orthonormal_form` -> `orthonormal_oneform`

(here and in similar comment below)